### PR TITLE
Fix snapshot wire order when simulation indices different than device wires

### LINKED ISSUE #6427: [BUG] Snapshot wires out of order
### Expected behavior

The order of the wires returned by a `qml.Snapshot()` should match those of the full execution

### Actual behavior

The wires are out of order, based on the order in which they had operations applied to them.

### Additional information

In the example I provide, I include a commented line that may be uncommented for an easy workaround, s.t. the snapshot's order matches that of the 'execution_result'

I found that a 2-wire example did not duplicate the issue, but returned the correct wire order in the snapshots.

I have not debugged these, but it seems that this may arise from a difference between lines 194 and 203 of `pennylane/debugging/snapshot.py` in the implementation of the measurements with a `wire_order`/`wire_map`

### Source code

```shell
@qml.qnode(qml.device("default.qubit", wires=3))
def circuit():
    #[qml.Identity(wires=x) for x in range(3)]
    qml.Hadamard(wires=2)
    qml.Snapshot("hadamard on 2")
    qml.Hadamard(wires=0)
    qml.Hadamard(wires=0)
    qml.Snapshot()
    return qml.state()
qml.snapshots(circuit)()
```

### Tracebacks

```shell
{'hadamard on 2': tensor([0.70710678+0.j, 0.        +0.j, 0.70710678+0.j, 0.        +0.j], requires_grad=True),
 1: tensor([0.70710678+0.j, 0.        +0.j, 0.70710678+0.j, 0.        +0.j], requires_grad=True),
 'execution_results': tensor([0.70710678+0.j, 0.70710678+0.j, 0.        +0.j, 0.        +0.j,
         0.        +0.j, 0.        +0.j, 0.        +0.j, 0.        +0.j], requires_grad=True)}
```

### System information

```shell
Name: PennyLane
Version: 0.38.0
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /home/jsmalley/qdrive/anaconda3/envs/penny2/lib/python3.9/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, packaging, pennylane-lightning, requests, rustworkx, scipy, toml, typing-extensions
Required-by: PennyLane-Catalyst, PennyLane-qiskit, PennyLane_Lightning

Platform info:           Linux-6.5.0-1025-azure-x86_64-with-glibc2.35
Python version:          3.9.18
Numpy version:           1.26.3
Scipy version:           1.11.4
Installed devices:
- default.clifford (PennyLane-0.38.0)
- default.gaussian (PennyLane-0.38.0)
- default.mixed (PennyLane-0.38.0)
- default.qubit (PennyLane-0.38.0)
- default.qubit.autograd (PennyLane-0.38.0)
- default.qubit.jax (PennyLane-0.38.0)
- default.qubit.legacy (PennyLane-0.38.0)
- default.qubit.tf (PennyLane-0.38.0)
- default.qubit.torch (PennyLane-0.38.0)
- default.qutrit (PennyLane-0.38.0)
- default.qutrit.mixed (PennyLane-0.38.0)
- default.tensor (PennyLane-0.38.0)
- null.qubit (PennyLane-0.38.0)
- lightning.qubit (PennyLane-Lightning-0.38.0)
- qiskit.aer (PennyLane-qiskit-0.36.0)
- qiskit.basicaer (PennyLane-qiskit-0.36.0)
- qiskit.basicsim (PennyLane-qiskit-0.36.0)
- qiskit.ibmq (PennyLane-qiskit-0.36.0)
- qiskit.ibmq.circuit_runner (PennyLane-qiskit-0.36.0)
- qiskit.ibmq.sampler (PennyLane-qiskit-0.36.0)
- qiskit.remote (PennyLane-qiskit-0.36.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

Internally, default qubit only performs the simulation with the wires that it needs for apply the operation.  Any wires only present in measurements are only added in at the very end.  This can be substantially more memory and time efficient in certain edge cases.

Unfortunately, this can cause confusion about the snapshotted state taken mid-simulation. Mid simulation we have fewer wires and a different wire order.

**Description of the Change:**

1) Add a `map_wires` method to snapshot to allow us to map the wires
2) Fill in device wires on Snapshots during `validate_device_wires`.
3) Allow `StateMP.process_state` to add in blank subsystems when the measurement has wires not present in the state's wire order.

These three steps are sufficient to make sure that snapshots always match the device's wire order.

**Benefits:**

Less Confusion

**Possible Drawbacks:**

It's no longer the state being used during the simulation at that point.  We are hiding away the fact we are working with a different state during the simulation.

**Related GitHub Issues:**

Fixes #6427 [sc-76515]

### CODE REVIEW

File: tests/measurements/test_state.py [hunk 1]
```diff
@@ -857,6 +871,8 @@ def func():
         elif len(return_wire_order) == 3:
             i, j, k = return_wire_order
             exp_statevector = np.kron(np.kron(single_states[i], single_states[j]), single_states[k])
+        else:
+            raise ValueError("how did I get here?")
```
Comment (albi3ro) [inline tests/measurements/test_state.py:875]: silencing pylint.

Comment (astralcai) [reply tests/measurements/test_state.py:875]: I feel that `assert` is typically more appropriate for branches that are never supposed to be reached.

File: doc/releases/changelog-0.39.0.md [hunk 2]
```diff
@@ -394,6 +394,9 @@
 
 <h3>Bug fixes 🐛</h3>
 
+* The wire order for `Snapshot`'s now matches the wire order of the device, rather than the simulation.
```
Comment (astralcai) [inline doc/releases/changelog-0.39.0.md:397]: Just out of curiosity, what is the expected behaviour if the device wasn't created with wires?

Comment (albi3ro) [reply doc/releases/changelog-0.39.0.md:397]: "Expected" can always be a tricky one, but I think if the device wasn't created with wires we have the flexibility to do things with whatever wire order we want.

**Review by lillian542 — APPROVED** [review]
LGTM!

File: pennylane/devices/preprocess.py [hunk 3]
```diff
@@ -149,16 +149,31 @@ def validate_device_wires(
...
-        measurements = tape.measurements.copy()
+
         modified = False
+        new_ops = None
+        for i, op in enumerate(tape.operations):
+            if isinstance(op, qml.Snapshot):
+                mp = op.hyperparameters["measurement"]
+                if not mp.wires:
+                    if not new_ops:
+                        new_ops = list(tape.operations)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/devices/preprocess.py:160]: Probably premature optimization, but I didn't want to copy the list of operations until I knew I needed to actually make a copy.

File: tests/measurements/test_state.py [hunk 4]
```diff
@@ -888,6 +904,8 @@ def func():
...
```
Comment (albi3ro) [inline tests/measurements/test_state.py:908]:
```suggestion
            exp_statevector = None # should not be reached
```

File: tests/measurements/test_state.py [hunk 5]
```diff
@@ -857,6 +871,8 @@ def func():
         elif len(return_wire_order) == 3:
             i, j, k = return_wire_order
             exp_statevector = np.kron(np.kron(single_states[i], single_states[j]), single_states[k])
+        else:
+            exp_statevector = None # should not be reached
```
Comment (astralcai) [inline tests/measurements/test_state.py:875]:
```suggestion
        else:  # len(return_wire_order) == 3
            i, j, k = return_wire_order
            exp_statevector = np.kron(np.kron(single_states[i], single_states[j]), single_states[k])
```

File: tests/measurements/test_state.py [hunk 6]
```diff
@@ -888,6 +904,8 @@ def func():
...
```
Comment (astralcai) [inline tests/measurements/test_state.py:908]:
```suggestion
        else:  # len(return_wire_order) == 3
            i, j, k = return_wire_order
            exp_statevector = np.kron(np.kron(single_states[i], single_states[j]), single_states[k])
```

**Review by astralcai — APPROVED** [review]
One suggestion, otherwise LGTM!

**Review by JerryChen97 — APPROVED** [review]

**Pull request merged**